source/agora/network/Crawler.d

```diff
@@ -0,0 +1,526 @@
+
+        Returns:
+            information about the network and about it's nodes
+
+
+    ***************************************************************************/
+
+
+    public CrawlResultHolder getNetworkInfo () @safe @nogc nothrow pure
+    {
+        return crawl_result_holder;
+    }
+
+    /***************************************************************************
+
+        Creates and returns a newly created NodeLocator object.
+
+        Returns:
+            a newly created NodeLocator object
+
+    ***************************************************************************/
+
+    protected INodeLocator makeNodeLocator ()
+    {
+        return new NodeLocatorGeoIP(this.config.node.ipdb_path);
+    }
+
+    // Start the Crawler
+    public void start ()
+    {
+        import std.array : array;
+
+        if (!this.config.node.collect_network_statistics)
+            return;
+
+        // at least 1 seed node needs to be provided
+        this.network_manager.seed_addresses.length || assert(0);
+
+        // node locator object should start successfully
+        this.node_locator.start() || assert(0);
+
+        this.processNetworkAddresses("seed", this.network_manager.seed_addresses[].array());
+
+        // start the crawling fibers
+        iota(config.node.num_of_crawlers).each!(_ =>
+            this.taskman.runTask(()
+            {
+                this.crawl();
+            })
+        );
+    }
+
+    // Initiate stopping the Crawler
+    public void stop () @safe @nogc nothrow pure
+    {
+        this.is_shutting_down = true;
+    }
+
+    // Stop the Crawler
+    private void do_stop ()
+    {
+        this.node_locator.stop();
+    }
+
+    /***************************************************************************
+
+        Processes network addresses by checking the addresses and updating the
+        internal data structures.
+
+        Params:
+            recommender_address = the node address which recommended `addresses`
+            addresses = the recommended addresses
+
+        Returns:
+            network clients that were created while processing the addresses
+
+    ***************************************************************************/
+
+    protected NetworkClientHolder[] processNetworkAddresses (Address recommender_address, Address[] addresses)
+    {
+        NetworkClientHolder[] network_client_holders;
+        network_client_holders.reserve(addresses.length);
+        foreach (address; addresses)
+            if (auto transformed_address = this.transformAddress(address))
+            {
+                auto network_client = this.getNetworkClient(transformed_address);
+                // Getting the client of an already shut down/never started node
+                // returns null in unittests if `use_non_assert_get_client` is
+                // set to `true`.
+                version (unittest)
+                    if (network_client is null)
+                        break;
```

Comment:
It's one of those special cases I was talking about. We need to solve this differently.